GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs

```diff
@@ -0,0 +1,87 @@
+        }
+
+        public override string Area => nameof(LooseObjectsStep);
+
+        protected override string LastRunTimeFilePath => Path.Combine(this.Context.Enlistment.GitObjectsRoot, "info", LooseObjectsLastRunFileName);
+        protected override TimeSpan TimeBetweenRuns => TimeSpan.FromDays(7);
+
+        public int CountLooseObjects()
+        {
+            int count = 0;
+            
+            foreach (string directoryPath in this.Context.FileSystem.EnumerateDirectories(this.Context.Enlistment.GitObjectsRoot))
+            {
+                string directoryName = directoryPath.TrimEnd(Path.DirectorySeparatorChar).Split(Path.DirectorySeparatorChar).Last();
+
+                if (SHA1Util.IsValidHex(directoryName, 2))
+                {                    
+                    count += this.Context.FileSystem.GetFiles(Path.Combine(this.Context.Enlistment.GitObjectsRoot, directoryPath), "*").Count();
+                }
+            }
+
+            return count;
+        }
+
+        protected override void PerformMaintenance()
+        {
+            using (ITracer activity = this.Context.Tracer.StartActivity(this.Area, EventLevel.Informational, Keywords.Telemetry, metadata: null))
+            {
+                try
+                {
+                    // forceRun is only currently tree for functional tests
+                    if (!this.forceRun)
+                    {
+                        if (!this.EnoughTimeBetweenRuns())
+                        {
+                            activity.RelatedWarning("Skipping step due to not enough time between runs");
+                            return;
+                        }
+
+                        IEnumerable<int> processIds = this.RunningGitProcessIds();
+                        if (processIds.Any())
+                        {
+                            activity.RelatedWarning("Skipping LooseObjectsStep due to git pids {0}", string.Join(",", processIds), Keywords.Telemetry);
+                            return;
+                        }
+                    }
+
+                    int beforeCount = this.CountLooseObjects();
+                    GitProcess.Result result = this.RunGitCommand((process) => process.PrunePacked(this.Context.Enlistment.GitObjectsRoot));
+                    int afterCount = this.CountLooseObjects();
+
+                    EventMetadata metadata = new EventMetadata();
+                    metadata.Add("StartingCount", beforeCount);
+                    metadata.Add("EndingCount", afterCount);
+                    metadata.Add("RemovedCount", beforeCount - afterCount);
+                    activity.RelatedEvent(EventLevel.Informational, this.Area, metadata, Keywords.Telemetry);
+
+                    this.SaveLastRunTimeToFile();
+                }
+                catch (Exception e)
+                {
+                    activity.RelatedError(this.CreateEventMetadata(e), "Failed to run LooseObjectsStep");

```

Comment:
Should this be a warning rather than an error?

We've typically reserved errors for scenarios where:

- The mount process is dying
- Processes interacting with VFS4G will see an I\O error

We've taken this approach so that if a customer sees a failure (typically build failure) they can scan the VFS4G logs to see if VFS4G is the likely the culprit of the failure (if there are errors, it's our fault).

Rather than using `RelatedError` would it be sufficient to use `RelatedWarning` + `Keywords.Telemetry`?

Reply:
It should be a warning, good call!